File: Submissions/687_Longest_Univalue_Path.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int longestUnivaluePath(TreeNode* root) {
        int result = 0;
        
        recursive(root, &result);
        
        return result;
    }
    
    int recursive(TreeNode* node, int* result) {
        if(node == nullptr) return 0;
        
        int left_length = recursive(node->left, result);
        int right_length = recursive(node->right, result);
        if(node->left && node->val == node->left->val){
            ++left_length;
        } else {
            left_length = 0;
        }
        
        if(node->right && node->val == node->right->val){
            ++right_length;
        } else {
            right_length = 0;
        }
        
        *result = std::max(*result, left_length + right_length);
        
        return std::max(left_length, right_length);
    }
};

File: Kotlin/687_Longest_Univalue_Path.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun longestUnivaluePath(root: TreeNode?): Int {
        if(root == null) return 0
        
        var result = maxOf(longestUnivaluePath(root.left), longestUnivaluePath(root.right))
        result = maxOf(result, recursive(root.left, root.`val`) + recursive(root.right, root.`val`))
        return result
    }
    
    fun recursive(node: TreeNode?, value: Int): Int {
        if(node == null || value != node.`val`) return 0
        
        val left = recursive(node.left, value)
        val right = recursive(node.right, value)
        return maxOf(left, right) + 1
    }
}